### DISCUSSION TITLE
How to properly make use of mail-encryption when Symfony is used by other frameworks?

### DISCUSSION THREAD

**Discussion by tantebootsy** [symfony/symfony#50748] [Q&A]
Hi there,
I'm working with the CMS TYPO3 which makes use of Symfony in some parts of its code e.g. when sending mails.

So the Mailer-class of the CMS `TYPO3\CMS\Core\Mail\MailMessage` extends `Symfony\Component\Mime\Email` which itself extends `Symfony\Component\Mime\Message`.

Now I want to encrypt a certain email which is sent by an extension of the CMS. For that I have to use an EventListener which hooks into the extension, encrypts the mail an then returns the mail back to the extension. So according to [the docs](https://symfony.com/doc/4.4/mailer.html#encrypting-messages) you have to feed the function `encrypt` from `Symfony\Component\Mime\Crypto\SMimeEncrypter` with `Symfony\Component\Mime\Email` which is fine because the EventDispatcher expects exactly that. But what the `encrypt` function does is returning `Symfony\Component\Mime\Message` which is think is rather strange. Because then I have the problem, that I cannot return this object back to the EventListener again.

I don't understand how to handle this situation and I don't understand why Symfony didn't take the old code from Swiftmail in which encrypting a message was simply done by a member of the Mail-object:

`
$smimeSigner = \Swift_Signers_SMimeSigner::newInstance();
$smimeSigner->setEncryptCertificate($encryptCertificateFile);
$email->attachSigner($smimeSigner);
`

S there, no new object of a different type was/is created but the original object ist altered – which I think is a cleaner approach, no? So is this a problem of how Symfony manages encryption or does the problem ly within the implementaion in the CMS? Or is it just me not seeing the obvious?

Any hints and explanations are highly appreciated. :)

Thx,
Michael

**Comment by stof** [comment]
to me, the fact that the Typo3 MailMessage class has a property holding the MailerInterface is a good sign that their code does not fit with the architecture of the Symfony component.

The mailer instance is a service that is responsible for sending emails. On the other hand, Message is a value object representing the Mime message for one email.
The Email class is a child class of Message providing many helper methods to make it easier to compose the message by mutating it. However, once the SMime encryption is done, attempting to use those helper would break things. That's why the SMimeEncrypter return a Message and not an Email. That's not an issue in Symfony because any email building must happen before performing the encryption. And the email sending in symfony/mailer expects a Message object (even a RawMessage object btw).

The Swiftmailer code would not fit in the symfony/mailer and symfony/mime architecture (the fact that the Swiftmailer code was mixing the concerns of configurable business logic and value representation was the main reason why we created a new mailer component from scratch in Symfony 5)

**Comment by tantebootsy** [comment]
Ok thx a lot for your explanation – that clears up things :) So my only option is to workaround the CMS and send the mail on my own?
